contentctl/objects/content_versioning_service.py

```diff
@@ -0,0 +1,530 @@
+        """
+        Using the cms_main index, validate content against the index to ensure our
+        savedsearches.conf is compatible with ES content versioning features. **NOTE**: while in
+        the future, this function may validate more types of content, currently, we only validate
+        detections against the cms_main index.
+        """
+        # Get the cached job and result count
+        result_count = self.get_num_cms_events(use_cache=True)
+        job = self._query_cms_main(use_cache=True)
+
+        # Create a running list of validation errors
+        exceptions: list[Exception] = []
+
+        # Generate an error for the count mismatch
+        if result_count != len(self.detections):
+            msg = (
+                f"[{self.infrastructure.instance_name}] Expected {len(self.detections)} matching "
+                f"events in cms_main, but found {result_count}."
+            )
+            self.logger.error(msg)
+            exceptions.append(Exception(msg))
+        self.logger.info(
+            f"[{self.infrastructure.instance_name}] Expecting {len(self.detections)} matching "
+            f"events in cms_main, found {result_count}."
+        )
+
+        # Init some counters and a mapping of detections to their names
+        count = 100
+        offset = 0
+        remaining_detections = {x.name: x for x in self.detections}
+        matched_detections: dict[str, Detection] = {}
+
+        # Create a filter for a specific memory error we're ok ignoring
+        sub_second_order_pattern = re.compile(
+            r".*Events might not be returned in sub-second order due to search memory limits.*"
+        )
+
+        # Iterate over the results until we've gone through them all
+        while offset < result_count:
+            iterator = ResultIterator(
+                response_reader=job.results(                                                        # type: ignore
+                    output_mode="json",
+                    count=count,
+                    offset=offset
+                ),
+                error_filters=[sub_second_order_pattern]
+            )
+
+            # Iterate over the currently fetched results
+            for cms_event in iterator:
+                # Increment the offset for each result
+                offset += 1
+
+                # Get the name of the search in the CMS event and attempt to use pattern matching
+                # to strip the prefix and suffix used for the savedsearches.conf name so we can
+                # compare to the detection
+                cms_entry_name = cms_event["action.correlationsearch.label"]
+                self.logger.info(
+                    f"[{self.infrastructure.instance_name}] {offset}: Matching cms_main entry "
+                    f"'{cms_entry_name}' against detections"
+                )
+                ptrn = re.compile(r"^" + self.global_config.app.label + r" - (?P<stripped_cms_entry_name>.+) - Rule$")
```

Comment:
I wonder if it's better to go in reverse now that we have the following function to generate a search name: 
https://github.com/splunk/contentctl/blob/98573b0fcc6040bc10046bf444c29d2685545ba6/contentctl/objects/abstract_security_content_objects/detection_abstract.py#L101

This also means if we change the template used in the link above to generate the detection name, we don't have to change it again in this function.

Reply:
Ahhhhh I don't think I knew about this function, I'll take a look!

Reply:
This one is _FAIRLY_ new. In the future, it might be a nice idea to actually associate the app config with a detection so that we don't have to carry/pass it around, making `action_dot_correlationsearch_dot_label` a cached_property instead of a function that takes an argument.